### PR TITLE
Jsonschema params/formdata endpoints

### LINKED ISSUE #309: Workflow Service endpoint for getting JSON schema
Add a `get_json_schema` endpoint to the workflow app template

### LINKED ISSUE #310: Workflow Service endpoint for validating JSON
Add a `validate_json` endpoint to the workflow app template, this would check the provided json against the workflow schema schema

### PR DESCRIPTION
**Pull request by cisaacstern**
closes #309
closes #310

### CODE REVIEW

**Comment by cisaacstern** [comment]
Next step:
- [ ] use datamodel code generator to add a formdata artifact to package
- [ ] add Params.from_formdata to "flatten" into Params (or, use a subclass whose only purpose is to make that translation with classmethod, so we don't have to touch datamodel code generator templating)

**Review by cisaacstern — COMMENTED** [review]
Trying out a new idea where I've reviewed my own PR below to add inline "field notes" as to what I've done here. But apart from the inline notes, the Big Idea™️ to grok from this PR is as follows:

- Workflows support task groups for grouping tasks for user readability/convenience. Task groups contain task instances with `id`s inside them.
- These groups are reflected in the RJSF jsonschema output (that is served from the `GET /params` endpoint added here). Therefore, the titles of the task groups become properties of the form data which is returned from a completed form.
- The workflows executor does not understand task groups, so before we hand off the config to the executor, we need to "flatten" it to remove the task groups from the form data. This is handled by the `POST /params` endpoint added here.

Which means a possible/recommended flow for configuring a workflow would look like this:

```mermaid
sequenceDiagram
    actor User
    participant Ecoscope Web
    participant Ecoscope Server
    participant Workflow Service
    participant Postgres

    User->>Ecoscope Web: "Configure a workflow please"
    Ecoscope Web->>Ecoscope Server: GET params for X workflow at Y version
    Ecoscope Server->>Workflow Service: GET /params
    Workflow Service-->>Ecoscope Server: rjsf jsonschema
    Ecoscope Server-->>Ecoscope Web: rjsf jsonschema
    Ecoscope Web-->>User: display Config Wizard
    User->>Ecoscope Web: fill in the form
    Ecoscope Web->>Ecoscope Server: POST /params (i.e. grouped formdata)
    Ecoscope Server->>Workflow Service: POST /params (i.e. grouped formadata)
    Workflow Service-->>Ecoscope Server: flat params
    Ecoscope Server->>Postgres: Store flat params
```

File: src/ecoscope-workflows-core/ecoscope_workflows_core/templates/tests/test_app.jinja2 [hunk 1]
```diff
@@ -28,3 +31,34 @@ def test_app(
...
+    assert invalid_request.status_code == 422
+
+    response = client.post("/params", json=formdata)
+    assert response.status_code == 200
+
+    if set(formdata) != set(case.params):
+        # this workflow uses task groups, so make a few extra asserts
+        # task groups are not required, so these asserts are skipped
+        # for workflows that simply use a flat layout
+        assert set(response.json()) == set(case.params)
```
Comment (cisaacstern) [inline src/ecoscope-workflows-core/ecoscope_workflows_core/templates/tests/test_app.jinja2:59]: Actually this assert can move out of the conditional I believe.

**Review by davidovichmarcos — APPROVED** [review]

File: src/ecoscope-workflows-core/ecoscope_workflows_core/templates/pkg/app.jinja2 [hunk 2]
```diff
@@ -103,3 +106,23 @@ def run(
...
+    formdata_asdict: dict[str, dict | Any] = formdata.model_dump()
+    params_fieldnames = Params.model_fields.keys()
+    params_kws = {}
+    for k, v in formdata_asdict.items():
+        if k in params_fieldnames:
+            params_kws[k] = v
+        else:
+            for inner_k, inner_v in v.items():
+                params_kws[inner_k] = inner_v
+    return Params(**params_kws)
```
[the remark this answers is not in the document]
Comment (cisaacstern) [reply src/ecoscope-workflows-core/ecoscope_workflows_core/templates/pkg/app.jinja2:128]: @juanlescano-ng I think we will have the `workflow_template_id`s based on 

Screenshot 2024-10-18 at 7 24 51 AM

something that may not be clear from this PR is that the frontend will not interact with the APIs added by this PR directly. we will be adding corresponding PRs on the `ecoscope-server` that are aware of workflow ids

[the remark this answers is not in the document]
Comment (cisaacstern) [reply src/ecoscope-workflows-core/ecoscope_workflows_core/templates/pkg/app.jinja2:128]: This is roughly captured in the sequence diagram in https://github.com/wildlife-dynamics/ecoscope-workflows/pull/326#pullrequestreview-2377361070 ... does that make sense?

[the remark this answers is not in the document]
Comment (juanlescano-ng) [reply src/ecoscope-workflows-core/ecoscope_workflows_core/templates/pkg/app.jinja2:128]: Gotcha!

**Review by juanlescano-ng — APPROVED** [review]
LGTM! 🚀

File: .github/workflows/examples.yml [hunk 3]
```diff
@@ -26,7 +26,33 @@ jobs:
...
 
-  test-examples:
+  test-examples-params:
+    needs: build-release
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        example: [
+          "patrol-events",
```
Comment (cisaacstern) [inline .github/workflows/examples.yml:36]:
```suggestion
          "events",
```
Rename for compatibility with #281 / #330

File: pixi.toml [hunk 4]
```diff
@@ -40,13 +39,18 @@ recompile-all-carryover-lockfile = { cmd ="echo 'recompile-all-carryover-lockfil
...
+pytest-patrols-params = { cmd = "./dev/pytest-examples-params.sh patrols" }
+pytest-subject-tracking-params = { cmd = "./dev/pytest-examples-params.sh subject-tracking" }
+
+pytest-patrol-events-app = { cmd = "./dev/pytest-examples-run.sh patrol-events 'app'" }
+pytest-patrols-app = { cmd = "./dev/pytest-examples-run.sh patrols 'app'" }
+pytest-subject-tracking-app = { cmd = "./dev/pytest-examples-run.sh subject-tracking 'app'" }
+
+pytest-patrol-events-cli = { cmd = "./dev/pytest-examples-run.sh patrol-events 'cli'" }
+pytest-patrols-cli = { cmd = "./dev/pytest-examples-run.sh patrols 'cli'" }
+pytest-subject-tracking-cli = { cmd = "./dev/pytest-examples-run.sh subject-tracking 'cli'" }
```
Comment (cisaacstern) [inline pixi.toml:53]:
```suggestion
pytest-events-app = { cmd = "./dev/pytest-examples-run.sh events 'app'" }
pytest-patrols-app = { cmd = "./dev/pytest-examples-run.sh patrols 'app'" }
pytest-subject-tracking-app = { cmd = "./dev/pytest-examples-run.sh subject-tracking 'app'" }

pytest-events-cli = { cmd = "./dev/pytest-examples-run.sh events 'cli'" }
pytest-patrols-cli = { cmd = "./dev/pytest-examples-run.sh patrols 'cli'" }
pytest-subject-tracking-cli = { cmd = "./dev/pytest-examples-run.sh subject-tracking 'cli'" }
```

File: src/ecoscope-workflows-core/ecoscope_workflows_core/templates/tests/test_app.jinja2 (hunk 1 above)
Comment (cisaacstern) [inline src/ecoscope-workflows-core/ecoscope_workflows_core/templates/tests/test_app.jinja2:59]:
```suggestion
    assert set(response.json()) == set(case.params)

    if set(formdata) != set(case.params):
        # this workflow uses task groups, so make one other assert
        # task groups are not required, so these asserts are skipped
        # for workflows that simply use a flat layout
```

File: pixi.toml [hunk 5]
```diff
@@ -40,13 +39,18 @@ recompile-all-carryover-lockfile = { cmd ="echo 'recompile-all-carryover-lockfil
...
 ]}
 
-pytest-events-app = { cmd = "./dev/pytest-examples.sh events 'app'" }
-pytest-patrols-app = { cmd = "./dev/pytest-examples.sh patrols 'app'" }
-pytest-subject-tracking-app = { cmd = "./dev/pytest-examples.sh subject-tracking 'app'" }
 
-pytest-events-cli = { cmd = "./dev/pytest-examples.sh events 'cli'" }
-pytest-patrols-cli = { cmd = "./dev/pytest-examples.sh patrols 'cli'" }
-pytest-subject-tracking-cli = { cmd = "./dev/pytest-examples.sh subject-tracking 'cli'" }
+pytest-patrol-events-params = { cmd = "./dev/pytest-examples-params.sh patrol-events" }
```
Comment (cisaacstern) [inline pixi.toml:43]:
```suggestion
pytest-events-params = { cmd = "./dev/pytest-examples-params.sh events" }
```

File: examples/events/test-cases.yaml [hunk 6]
```diff
@@ -117,11 +117,6 @@ test1:
       title: "Density Map" # (<class 'str'>, FieldInfo(annotation=NoneType, required=True, description='The title of the widget'))
     # Parameters for 'Split Events by Group' using task `split_groups`.
     split_event_groups: {}
-    # Parameters for 'Grouped Events Colormap' using task `apply_color_map`.
-    grouped_events_colormap:
-      input_column_name: "event_type" # (<class 'str'>, FieldInfo(annotation=NoneType, required=True, description='The name of the column with categorical values.'))
-      colormap: "tab20b" # (str | list[str], FieldInfo(annotation=NoneType, required=True, description='Either a named mpl.colormap or a list of string hex values.'))
-      output_column_name: "event_type_colormap"
```
Comment (cisaacstern) [inline examples/events/test-cases.yaml:124]: This task instance does not appear to be used at all in the spec, so removing these params.

**Pull request merged**